storage/aptosdb/src/state_store/mod.rs

```diff
@@ -518,7 +531,36 @@ impl StateStore {
 
         let base_version = first_version.checked_sub(1);
         let mut usage = self.get_usage(base_version)?;
-        let mut cache = HashMap::<StateKey, (Version, Option<StateValue>)>::new();
+        let cache = Arc::new(Mutex::new(
+            HashMap::<StateKey, (Version, Option<StateValue>)>::new(),
+        ));
+
+        if let Some(base_version) = base_version {
+            let key_set = value_state_sets
+                .iter()
+                .flat_map(|value_state_set| value_state_set.iter())
+                .map(|(key, _)| key)
+                .collect::<HashSet<_>>();
+            IO_POOL.scope(|s| {
+                for key in key_set {
+                    let cache = cache.clone();
+                    s.spawn(move |_| {
+                        let _timer = OTHER_TIMERS_SECONDS
+                            .with_label_values(&["put_stats_and_indices__get_state_value"])
+                            .start_timer();
+                        let version_and_value = self
+                            .state_db
+                            .get_state_value_with_version_by_version(key, base_version)
+                            .expect("Must succeed.");
+                        if let Some((version, value)) = version_and_value {
+                            cache.lock().insert(key.clone(), (version, Some(value)));
```

Comment:
Use a concurrent hash map here instead of locking?

Reply:
https://github.com/xacrimon/dashmap- This is something we seem to be using in other places for lock free hashmap.

Reply:
Dashmap is slower at least when we tried to make SchemaBatch concurrent -- not a bottleneck, I won't even bother to try again here.

Reply:
I remember @msmouse said the dashmap is not that good, worse than his sharded hash map..

Given this part is not the bottleneck after this change, I'd prefer to leave it as is, and do some investigation first.

Reply:
@grao1991  in the case of SchemaBatch it's slower than a plain Mutex<HashMap>, not even sharded.